mat/dense_arithmetic.go

```diff
@@ -457,67 +457,162 @@ func strictCopy(m *Dense, a Matrix) {
 
 // Exp calculates the exponential of the matrix a, e^a, placing the result
 // in the receiver. Exp will panic with matrix.ErrShape if a is not square.
-//
-// Exp uses the scaling and squaring method described in section 3 of
-// http://www.cs.cornell.edu/cv/researchpdf/19ways+.pdf.
 func (m *Dense) Exp(a Matrix) {
+	// The implementation used here is from Functions of Matrices: Theory and Computation
+	// Chapter 10, Algorithm 10.20. https://doi.org/10.1137/1.9780898717778.ch10
+
 	r, c := a.Dims()
 	if r != c {
 		panic(ErrShape)
 	}
 
-	var w *Dense
-	if m.IsZero() {
-		m.reuseAsZeroed(r, r)
-		w = m
-	} else {
-		w = getWorkspace(r, r, true)
-	}
-	for i := 0; i < r*r; i += r + 1 {
-		w.mat.Data[i] = 1
+	m.reuseAs(r, r)
+	if r == 1 {
+		m.mat.Data[0] = math.Exp(a.At(0, 0))
+		return
 	}
 
-	const (
-		terms   = 10
-		scaling = 4
-	)
-
-	small := getWorkspace(r, r, false)
-	small.Scale(math.Pow(2, -scaling), a)
-	power := getWorkspace(r, r, false)
-	power.Copy(small)
+	pade := []struct {
+		phi float64
+		b   []float64
+	}{
+		{phi: 0.015, b: []float64{
+			120, 60, 12, 1,
+		}},
+		{phi: 0.25, b: []float64{
+			30240, 15120, 3360, 420, 30, 1,
+		}},
+		{phi: 0.95, b: []float64{
+			17297280, 8648640, 1995840, 277200, 25200, 1512, 56, 1,
+		}},
+		{phi: 2.1, b: []float64{
+			17643225600, 8821612800, 2075673600, 302702400, 30270240, 2162160, 110880, 3960, 90, 1,
+		}},
+	}
+
+	a1 := m
+	a1.Copy(a)
+	v := getWorkspace(r, r, true)
+	vraw := v.RawMatrix()
+	vvec := blas64.Vector{Inc: 1, Data: vraw.Data}
+	defer putWorkspace(v)
+
+	u := getWorkspace(r, r, true)
+	uraw := u.RawMatrix()
+	uvec := blas64.Vector{Inc: 1, Data: uraw.Data}
+	defer putWorkspace(u)
+
+	a2 := getWorkspace(r, r, false)
+	defer putWorkspace(a2)
+
+	n1 := Norm(a, 1)
+	for i, t := range pade {
+		if n1 > t.phi {
+			continue
+		}
+
+		// This loop only executes once, so
+		// this is not as horrible as it looks.
+		p := getWorkspace(r, r, true)
+		praw := p.RawMatrix()
+		pvec := blas64.Vector{Inc: 1, Data: praw.Data}
+		defer putWorkspace(p)
 
-	var (
-		tmp   = getWorkspace(r, r, false)
-		factI = 1.
-	)
-	for i := 1.; i < terms; i++ {
-		factI *= i
-
-		// This is OK to do because power and tmp are
-		// new Dense values so all rows are contiguous.
-		// TODO(kortschak) Make this explicit in the NewDense doc comment.
-		for j, v := range power.mat.Data {
-			tmp.mat.Data[j] = v / factI
+		for i := 0; i < r; i++ {
```

Comment:
Could you use k here so it's obviously different than the outer loop i?

Reply:
Done.